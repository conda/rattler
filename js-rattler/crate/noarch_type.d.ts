/**
 * Noarch packages are packages that are not architecture specific and therefore
 * only have to be built once. A `NoArchType` is either specific to an
 * architecture or not.
 *
 * @public
 */
export declare type NoArchType = undefined | true | "python" | "generic";
