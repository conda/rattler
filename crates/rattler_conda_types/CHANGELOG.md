# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.27.3](https://github.com/conda/rattler/compare/rattler_conda_types-v0.27.2...rattler_conda_types-v0.27.3) - 2024-09-02

### Added
- add edge case tests for `StringMatcher` ([#839](https://github.com/conda/rattler/pull/839))

## [0.27.2](https://github.com/conda/rattler/compare/rattler_conda_types-v0.27.1...rattler_conda_types-v0.27.2) - 2024-08-15

### Added
- add extra field ([#811](https://github.com/conda/rattler/pull/811))
- parse `channel` key and consolidate `NamelessMatchSpec` ([#810](https://github.com/conda/rattler/pull/810))

### Fixed
- move more links to the conda org from conda-incubator ([#816](https://github.com/conda/rattler/pull/816))
- use conda-incubator

### Other
- change links from conda-incubator to conda ([#813](https://github.com/conda/rattler/pull/813))
- update banner ([#808](https://github.com/conda/rattler/pull/808))

## [0.27.1](https://github.com/baszalmstra/rattler/compare/rattler_conda_types-v0.27.0...rattler_conda_types-v0.27.1) - 2024-08-06

### Fixed
- parse `~=` as version not as path ([#804](https://github.com/baszalmstra/rattler/pull/804))

## [0.27.0](https://github.com/baszalmstra/rattler/compare/rattler_conda_types-v0.26.3...rattler_conda_types-v0.27.0) - 2024-08-02

### Fixed
- redact secrets in the `canonical_name` functions ([#801](https://github.com/baszalmstra/rattler/pull/801))
- make `base_url` of `Channel` always contain a trailing slash ([#800](https://github.com/baszalmstra/rattler/pull/800))
- parse channel in matchspec string ([#792](https://github.com/baszalmstra/rattler/pull/792))
- constraints on virtual packages were ignored ([#795](https://github.com/baszalmstra/rattler/pull/795))
- url parsing for namelessmatchspec and cleanup functions ([#790](https://github.com/baszalmstra/rattler/pull/790))

### Other
- mark some crates 1.0 ([#789](https://github.com/baszalmstra/rattler/pull/789))

## [0.26.3](https://github.com/conda/rattler/compare/rattler_conda_types-v0.26.2...rattler_conda_types-v0.26.3) - 2024-07-23

### Fixed
- channel `base_url` requires trailing slash ([#787](https://github.com/conda/rattler/pull/787))

## [0.26.2](https://github.com/conda/rattler/compare/rattler_conda_types-v0.26.1...rattler_conda_types-v0.26.2) - 2024-07-23

### Added
- `environment.yaml` type ([#786](https://github.com/conda/rattler/pull/786))
- Add to_path() method to ExplicitEnvironmentSpec ([#781](https://github.com/conda/rattler/pull/781))
- expose `HasPrefixEntry` for public use ([#784](https://github.com/conda/rattler/pull/784))

## [0.26.1](https://github.com/conda/rattler/compare/rattler_conda_types-v0.26.0...rattler_conda_types-v0.26.1) - 2024-07-15

### Other
- PrefixRecord deserialization using simd ([#777](https://github.com/conda/rattler/pull/777))

## [0.26.0](https://github.com/conda/rattler/compare/rattler_conda_types-v0.25.2...rattler_conda_types-v0.26.0) - 2024-07-08

### Added
- add support for zos-z ([#753](https://github.com/conda/rattler/pull/753))
- return pybytes for sha256 and md5 everywhere and use md5 hash for legacy bz2 md5 ([#752](https://github.com/conda/rattler/pull/752))
- add direct url repodata building ([#725](https://github.com/conda/rattler/pull/725))
- add shards_base_url and write shards atomically ([#747](https://github.com/conda/rattler/pull/747))

### Fixed
- allow version following package in strict mode ([#770](https://github.com/conda/rattler/pull/770))
- Fix doctests and start testing them again ([#767](https://github.com/conda/rattler/pull/767))
- skip over implicit `0` components when copying ([#760](https://github.com/conda/rattler/pull/760))
- allow empty json repodata ([#745](https://github.com/conda/rattler/pull/745))
- lenient and strict parsing of equality signs ([#738](https://github.com/conda/rattler/pull/738))
- This fixes parsing of `ray[default,data] >=2.9.0,<3.0.0` ([#732](https://github.com/conda/rattler/pull/732))

## [0.25.2](https://github.com/baszalmstra/rattler/compare/rattler_conda_types-v0.25.1...rattler_conda_types-v0.25.2) - 2024-06-04

### Added
- parse url and path as matchspec ([#704](https://github.com/baszalmstra/rattler/pull/704))

### Fixed
- issue 722 ([#723](https://github.com/baszalmstra/rattler/pull/723))

### Other
- remove lfs ([#512](https://github.com/baszalmstra/rattler/pull/512))
- move the cache tooling into its own crate for reuse downstream ([#721](https://github.com/baszalmstra/rattler/pull/721))

## [0.25.1](https://github.com/conda/rattler/compare/rattler_conda_types-v0.25.0...rattler_conda_types-v0.25.1) - 2024-06-03

### Added
- add a `with_alpha` function that adds `0a0` to the version ([#696](https://github.com/conda/rattler/pull/696))

## [0.25.0](https://github.com/conda/rattler/compare/rattler_conda_types-v0.24.0...rattler_conda_types-v0.25.0) - 2024-05-28

### Added
- when bumping, extend versions with `0` to match the bump request ([#695](https://github.com/conda/rattler/pull/695))
- extend tests and handle characters better when bumping versions ([#694](https://github.com/conda/rattler/pull/694))
- add a function to extend version with `0s` ([#689](https://github.com/conda/rattler/pull/689))
- add run exports to package data ([#671](https://github.com/conda/rattler/pull/671))

### Fixed
- lenient parsing of 2023.*.* ([#688](https://github.com/conda/rattler/pull/688))
- VersionSpec starts with, with trailing zeros ([#686](https://github.com/conda/rattler/pull/686))

### Other
- move bump implementation to bump.rs and simplify tests ([#692](https://github.com/conda/rattler/pull/692))

## [0.24.0](https://github.com/conda/rattler/compare/rattler_conda_types-v0.23.1...rattler_conda_types-v0.24.0) - 2024-05-27

### Added
- removed Ord and more ([#673](https://github.com/conda/rattler/pull/673))
- always store purls as a key in lock file ([#669](https://github.com/conda/rattler/pull/669))
- add solve strategies ([#660](https://github.com/conda/rattler/pull/660))

### Fixed
- make topological sorting support fully cyclic dependencies ([#678](https://github.com/conda/rattler/pull/678))

## [0.23.1](https://github.com/conda/rattler/compare/rattler_conda_types-v0.23.0...rattler_conda_types-v0.23.1) - 2024-05-14

### Added
- exclude repodata records based on timestamp ([#654](https://github.com/conda/rattler/pull/654))

## [0.23.0](https://github.com/conda/rattler/compare/rattler_conda_types-v0.22.1...rattler_conda_types-v0.23.0) - 2024-05-13

### Added
- high level repodata access ([#560](https://github.com/conda/rattler/pull/560))

### Other
- update README.md

## [0.22.1](https://github.com/conda/rattler/compare/rattler_conda_types-v0.22.0...rattler_conda_types-v0.22.1) - 2024-05-06

### Added
- expose `*Record.noarch` in Python bindings ([#635](https://github.com/conda/rattler/pull/635))

## [0.22.0](https://github.com/conda/rattler/compare/rattler_conda_types-v0.21.0...rattler_conda_types-v0.22.0) - 2024-04-25

### Added
- add support for extracting prefix placeholder data to PathsEntry ([#614](https://github.com/conda/rattler/pull/614))

## [0.21.0](https://github.com/conda/rattler/compare/rattler_conda_types-v0.20.5...rattler_conda_types-v0.21.0) - 2024-04-19

### Added
- make root dir configurable in channel config ([#602](https://github.com/conda/rattler/pull/602))

### Fixed
- better value for `link` field ([#610](https://github.com/conda/rattler/pull/610))

### Other
- update dependencies incl. reqwest ([#606](https://github.com/conda/rattler/pull/606))

## [0.20.5](https://github.com/baszalmstra/rattler/compare/rattler_conda_types-v0.20.4...rattler_conda_types-v0.20.5) - 2024-04-05

### Fixed
- run post-link scripts ([#574](https://github.com/baszalmstra/rattler/pull/574))

## [0.20.4](https://github.com/conda/rattler/compare/rattler_conda_types-v0.20.3...rattler_conda_types-v0.20.4) - 2024-03-30

### Fixed
- matchspec empty namespace and channel cannonical name ([#582](https://github.com/conda/rattler/pull/582))

## [0.20.3](https://github.com/conda/rattler/compare/rattler_conda_types-v0.20.2...rattler_conda_types-v0.20.3) - 2024-03-21

### Fixed
- allow not starts with in strict mode ([#577](https://github.com/conda/rattler/pull/577))

## [0.20.2](https://github.com/conda/rattler/compare/rattler_conda_types-v0.20.1...rattler_conda_types-v0.20.2) - 2024-03-14

### Other
- add pixi badge ([#563](https://github.com/conda/rattler/pull/563))

## [0.20.1](https://github.com/conda/rattler/compare/rattler_conda_types-v0.20.0...rattler_conda_types-v0.20.1) - 2024-03-08

### Fixed
- chrono deprecation warnings ([#558](https://github.com/conda/rattler/pull/558))

## [0.20.0](https://github.com/conda/rattler/compare/rattler_conda_types-v0.19.0...rattler_conda_types-v0.20.0) - 2024-03-06

### Added
- [**breaking**] optional strict parsing of matchspec and versionspec ([#552](https://github.com/conda/rattler/pull/552))

### Fixed
- patch unsupported glob operators ([#551](https://github.com/conda/rattler/pull/551))
- dont use workspace dependencies for local crates ([#546](https://github.com/conda/rattler/pull/546))

### Other
- every crate should have its own version ([#557](https://github.com/conda/rattler/pull/557))

## [0.19.0](https://github.com/baszalmstra/rattler/compare/rattler_conda_types-v0.18.0...rattler_conda_types-v0.19.0) - 2024-02-26

### Fixed
- Fix arch for osx-arm64 and win-arm64 ([#528](https://github.com/baszalmstra/rattler/pull/528))
- Channel name display ([#531](https://github.com/baszalmstra/rattler/pull/531))
