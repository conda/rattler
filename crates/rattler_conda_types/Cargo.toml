[package]
name = "rattler_conda_types"
version = "0.27.2"
edition.workspace = true
authors = ["Bas Zalmstra <zalmstra.bas@gmail.com>"]
description = "Rust data types for common types used within the Conda ecosystem"
categories.workspace = true
homepage.workspace = true
repository.workspace = true
license.workspace = true
readme.workspace = true

[dependencies]
chrono = { workspace = true }
file_url = { path = "../file_url", version = "0.1.4" }
fxhash = { workspace = true }
glob = { workspace = true }
hex = { workspace = true }
itertools = { workspace = true }
lazy-regex = { workspace = true }
nom = { workspace = true }
purl = { workspace = true, features = ["serde"] }
rattler_digest = { path = "../rattler_digest", version = "1.0.1", default-features = false, features = ["serde"] }
rattler_macros = { path = "../rattler_macros", version = "1.0.1", default-features = false }
regex = { workspace = true }
simd-json = { workspace = true , features = ["serde_impl"]}
serde = { workspace = true, features = ["derive", "rc"] }
serde_json = { workspace = true }
serde_repr = { workspace = true }
serde_with = { workspace = true, features = ["indexmap_2"] }
serde-untagged = { workspace = true }
serde_yaml = { workspace = true }
smallvec = { workspace = true, features = ["serde", "const_new", "const_generics", "union"] }
strum = { workspace = true, features = ["derive"] }
thiserror = { workspace = true }
tracing = { workspace = true }
typed-path = { workspace = true }
url = { workspace = true, features = ["serde"] }
indexmap = { workspace = true }
rattler_redaction = { version = "0.1.1", path = "../rattler_redaction" }
dirs = { workspace = true }

[dev-dependencies]
rand = { workspace = true }
insta = { workspace = true, features = ["yaml", "redactions", "toml", "glob", "filters"] }
rattler_package_streaming = { path = "../rattler_package_streaming", default-features = false, features = ["rustls-tls"] }
tempfile = { workspace = true }
rstest = { workspace = true }
assert_matches = { workspace = true }
hex-literal = { workspace = true }
criterion = { workspace = true, features = ["html_reports"] }
pathdiff = { workspace = true }
dunce = { workspace = true }
tools = { path = "../tools" }

[[bench]]
name = "parse"
harness = false

[[bench]]
name = "prefix_record_from_path"
harness = false
