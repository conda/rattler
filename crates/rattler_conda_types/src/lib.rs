#![deny(missing_docs)]
//! `rattler-conda-types` contains data models for types commonly found within
//! the Conda ecosystem. The library itself doesnt provide any functionality
//! besides parsing the data types.

mod build_spec;
mod channel;
mod channel_data;
mod explicit_environment_spec;
mod match_spec;
mod no_arch_type;
mod parse_mode;
mod platform;
mod repo_data;
mod repo_data_record;
mod run_export;
mod utils;
mod version;
pub mod version_spec;

mod environment_yaml;
mod generic_virtual_package;
pub mod package;
mod package_name;
pub mod prefix_record;

#[cfg(test)]
use std::path::{Path, PathBuf};

pub use build_spec::{BuildNumber, BuildNumberSpec, ParseBuildNumberSpecError};
pub use channel::{Channel, ChannelConfig, ChannelUrl, NamedChannelOrUrl, ParseChannelError};
pub use channel_data::{ChannelData, ChannelDataPackage};
pub use environment_yaml::{EnvironmentYaml, MatchSpecOrSubSection};
pub use explicit_environment_spec::{
    ExplicitEnvironmentEntry, ExplicitEnvironmentSpec, PackageArchiveHash,
    ParseExplicitEnvironmentSpecError, ParsePackageArchiveHashError,
};
pub use generic_virtual_package::GenericVirtualPackage;
pub use match_spec::{
    matcher::{StringMatcher, StringMatcherParseError},
    parse::ParseMatchSpecError,
    MatchSpec, Matches, NamelessMatchSpec,
};
pub use no_arch_type::{NoArchKind, NoArchType};
pub use package_name::{InvalidPackageNameError, PackageName};
pub use parse_mode::ParseStrictness;
pub use platform::{Arch, ParseArchError, ParsePlatformError, Platform};
pub use prefix_record::PrefixRecord;
pub use repo_data::{
    compute_package_url,
    patches::{PackageRecordPatch, PatchInstructions, RepoDataPatch},
    sharded::{Shard, ShardedRepodata, ShardedSubdirInfo},
    ChannelInfo, ConvertSubdirError, PackageRecord, RepoData, ValidatePackageRecordsError,
};
pub use repo_data_record::RepoDataRecord;
pub use run_export::RunExportKind;
pub use version::{
    Component, ParseVersionError, ParseVersionErrorKind, StrictVersion, Version, VersionBumpError,
    VersionBumpType, VersionExtendError, VersionWithSource,
};
pub use version_spec::VersionSpec;

/// An package identifier that can be used to identify packages across package
/// ecosystems.
pub type PackageUrl = purl::GenericPurl<String>;

#[cfg(test)]
pub(crate) fn get_test_data_dir() -> PathBuf {
    Path::new(env!("CARGO_MANIFEST_DIR")).join("../../test-data")
}
