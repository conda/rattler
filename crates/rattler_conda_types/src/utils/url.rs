/// Parses the schema part of the human-readable channel. Returns the scheme part if it exists.
pub(crate) fn parse_scheme(channel: &str) -> Option<&str> {
    let scheme_end = channel.find("://")?;

    // Scheme part is too long
    if scheme_end > 11 {
        return None;
    }

    let scheme_part = &channel[0..scheme_end];
    let mut scheme_chars = scheme_part.chars();

    // First character must be alphabetic
    if scheme_chars.next().map(char::is_alphabetic) != Some(true) {
        return None;
    }

    // The rest must be alpha-numeric
    if scheme_chars.all(char::is_alphanumeric) {
        Some(scheme_part)
    } else {
        None
    }
}
