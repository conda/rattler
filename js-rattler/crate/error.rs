use rattler_conda_types::version_spec::ParseVersionSpecError;
use rattler_conda_types::{ParseVersionError, VersionBumpError, VersionExtendError};
use thiserror::Error;
use wasm_bindgen::JsValue;

#[derive(Debug, Error)]
pub enum JsError {
    #[error(transparent)]
    InvalidVersion(#[from] ParseVersionError),
    #[error(transparent)]
    VersionExtendError(#[from] VersionExtendError),
    #[error(transparent)]
    VersionBumpError(#[from] VersionBumpError),

    #[error(transparent)]
    ParseVersionSpecError(#[from] ParseVersionSpecError),
}

pub type JsResult<T> = Result<T, JsError>;

impl From<JsError> for JsValue {
    fn from(error: JsError) -> Self {
        match error {
            JsError::InvalidVersion(error) => JsValue::from_str(&error.to_string()),
            JsError::VersionExtendError(error) => JsValue::from_str(&error.to_string()),
            JsError::VersionBumpError(error) => JsValue::from_str(&error.to_string()),
            JsError::ParseVersionSpecError(error) => JsValue::from_str(&error.to_string()),
        }
    }
}
