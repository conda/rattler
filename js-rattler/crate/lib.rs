mod error;
mod gateway;
pub mod solve;
mod utils;
mod version;
mod version_spec;

pub use error::{JsError, JsResult};
use rattler_conda_types::ParseStrictness;

use wasm_bindgen::prelude::*;

// When the `wee_alloc` feature is enabled, use `wee_alloc` as the global
// allocator.
#[cfg(feature = "wee_alloc")]
#[global_allocator]
static ALLOC: wee_alloc::WeeAlloc = wee_alloc::WeeAlloc::INIT;

/// This function is called when the wasm module is instantiated.
#[wasm_bindgen(start)]
pub fn start() {
    utils::set_panic_hook();
}

/// Defines how strict a parser should be when parsing an object from a string.
/// @public
#[wasm_bindgen(js_name=ParseStrictness)]
pub enum JsParseStrictness {
    Strict,
    Lenient,
}

impl From<ParseStrictness> for JsParseStrictness {
    fn from(value: ParseStrictness) -> Self {
        match value {
            ParseStrictness::Strict => JsParseStrictness::Strict,
            ParseStrictness::Lenient => JsParseStrictness::Lenient,
        }
    }
}

impl From<JsParseStrictness> for ParseStrictness {
    fn from(value: JsParseStrictness) -> Self {
        match value {
            JsParseStrictness::Strict => ParseStrictness::Strict,
            JsParseStrictness::Lenient => ParseStrictness::Lenient,
        }
    }
}
