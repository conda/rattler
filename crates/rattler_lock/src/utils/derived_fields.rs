//! This module contains functions to derive certain fields from other fields in
//! a [`CondaPackageData`].
//!
//! For instance if we know the location of a package, and it contains a valid
//! archive identifier filename, then we can derive the name, version, build,
//! subdir and channel from it (see [`LocationDerivedFields`]).
//!
//! Care must be taken when changing the logic in this module, as it encodes how
//! we resolve empty fields in a lock-file. This is important for
//! reproducibility.

use std::fmt::{Debug, Formatter};
use std::str::FromStr;

use rattler_conda_types::{
    package::ArchiveIdentifier, BuildNumber, ChannelUrl, NoArchType, PackageName, Platform,
    VersionWithSource,
};
use url::Url;

use crate::UrlOrPath;

/// A helper struct that wraps all fields of a [`CondaPackageData`] that can be
/// derived just from the location of the package.
pub(crate) struct LocationDerivedFields {
    pub file_name: Option<String>,
    pub name: Option<PackageName>,
    pub version: Option<VersionWithSource>,
    pub build: Option<String>,
    pub subdir: Option<String>,
    pub channel: Option<ChannelUrl>,
}

impl Debug for LocationDerivedFields {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("LocationDerivedFields")
            .field("file_name", &self.file_name)
            .field("name", &self.name.as_ref().map(PackageName::as_source))
            .field("version", &self.version.as_ref().map(|s| s.as_str()))
            .field("build", &self.build)
            .field("subdir", &self.subdir)
            .field("channel", &self.channel.as_ref().map(ChannelUrl::as_str))
            .finish()
    }
}

impl LocationDerivedFields {
    /// Constructs a new instance by deriving all fields from the given
    /// location.
    pub fn new(location: &UrlOrPath) -> Self {
        let (file_name, archive_identifier) = location
            .file_name()
            .and_then(|f| ArchiveIdentifier::try_from_filename(f).map(|a| (f.to_string(), a)))
            .map_or((None, None), |(f, a)| (Some(f), Some(a)));
        let (name, version, build) = archive_identifier
            .and_then(|a| {
                Some((
                    PackageName::new_unchecked(a.name),
                    VersionWithSource::from_str(&a.version).ok()?,
                    a.build_string,
                ))
            })
            .map_or((None, None, None), |(name, version, build_string)| {
                (Some(name), Some(version), Some(build_string))
            });
        let subdir = derive_subdir_from_location(location).map(ToString::to_string);
        let channel = derive_channel_from_location(location);
        Self {
            file_name,
            name,
            version,
            build,
            subdir,
            channel,
        }
    }
}

/// Try to derive the build number from the build string. It is common to append
/// the build number to the end of the build string.
pub(crate) fn derive_build_number_from_build(build: &str) -> Option<BuildNumber> {
    let (_, trailing_number_str) = build
        .rsplit_once(|c: char| !c.is_ascii_digit())
        .unwrap_or(("", build));

    trailing_number_str.parse().ok()
}

/// Try to derive the subdir from a common conda URL. This assumes that the URL
/// is formatted as `scheme://../subdir/name-version-build.ext`.
pub(crate) fn derive_subdir_from_location(location: &UrlOrPath) -> Option<&str> {
    match location {
        UrlOrPath::Url(url) => derive_subdir_from_url(url),
        UrlOrPath::Path(_) => None,
    }
}

/// Try to derive the subdir from a common conda URL. This assumes that the URL
/// is formatted as `scheme://../subdir/name-version-build.ext`.
pub fn derive_subdir_from_url(url: &Url) -> Option<&str> {
    if !(url.scheme() == "file" || url.scheme() == "http" || url.scheme() == "https") {
        return None;
    }

    let mut path_iter = url.path_segments()?.rev();
    let archive_str = path_iter.next()?;
    let subdir_str = path_iter.next()?;

    // Try to parse the archive string as an archive identifier. If it fails we
    // can't derive the subdir.
    let _ = ArchiveIdentifier::try_from_filename(archive_str)?;

    // Parse the subdir as a platform, if it fails we can't derive the subdir.
    Platform::from_str(subdir_str).is_ok().then_some(subdir_str)
}

/// Channel from url, this is everything before the filename and the subdir
/// So for example: <https://conda.anaconda.org/conda-forge/> is a channel name
/// that we parse from something like: <https://conda.anaconda.org/conda-forge/osx-64/python-3.11.0-h4150a38_1_cpython.conda>
pub(crate) fn derive_channel_from_url(url: &Url) -> Option<ChannelUrl> {
    let mut result = url.clone();

    // Strip the last two path segments. We assume the first one contains the
    // file_name, and the other the subdirectory.
    result.path_segments_mut().ok()?.pop().pop();

    Some(result.into())
}

/// Returns the channel when deriving it from the location if possible.
pub(crate) fn derive_channel_from_location(url: &UrlOrPath) -> Option<ChannelUrl> {
    match url {
        UrlOrPath::Url(url) => derive_channel_from_url(url),
        UrlOrPath::Path(_) => None,
    }
}

pub(crate) fn derive_arch_and_platform(subdir: &str) -> (Option<String>, Option<String>) {
    let platform = Platform::from_str(subdir).ok();
    platform.map_or((None, None), |p| {
        (
            p.arch().map(|arch| arch.to_string()),
            p.only_platform().map(ToString::to_string),
        )
    })
}

/// Derive the noarch type from the subdir and the build string.
///
/// By default, the noarch type is set to `None`. However, if the channel is
/// `noarch`, then we assume that the package is a noarch package but it does
/// not tell us whether this is a generic or python noarch package. As a
/// heuristic we then look at the build string, if it contains the word `py` we
/// assume that this is a python noarch package.
pub(crate) fn derive_noarch_type(subdir: &str, build_string: &str) -> NoArchType {
    let is_python_build_string = build_string.contains("py");
    match subdir {
        "noarch" => {
            if is_python_build_string {
                NoArchType::python()
            } else {
                NoArchType::generic()
            }
        }
        _ => NoArchType::none(),
    }
}

#[cfg(test)]
mod test {
    use super::*;
    use rstest::*;

    #[test]
    fn test_derive_build_number_from_build() {
        assert_eq!(derive_build_number_from_build("2"), Some(2));
        assert_eq!(derive_build_number_from_build("1.2.3"), Some(3));
        assert_eq!(derive_build_number_from_build("1.2.3-4"), Some(4));
        assert_eq!(derive_build_number_from_build("py313hb6a6212_1"), Some(1));
        assert_eq!(derive_build_number_from_build("foobar"), None);
        assert_eq!(derive_build_number_from_build("123_"), None);
        assert_eq!(derive_build_number_from_build("123foo"), None);
        assert_eq!(derive_build_number_from_build("pyhd8ed1ab_100"), Some(100));
        assert_eq!(derive_build_number_from_build("h803f02a_3"), Some(3));
    }

    #[test]
    fn test_derive_subdir_from_url() {
        assert_eq!(
            derive_subdir_from_url(
                &"https://conda.anaconda.org/conda-forge/linux-aarch64/qt-5.15.8-h803f02a_0.conda"
                    .parse()
                    .unwrap()
            ),
            Some("linux-aarch64")
        );
        assert_eq!(
            derive_subdir_from_url(
                &"https://conda.anaconda.org/conda-forge/osx-arm64/x264-1!164.3095-h57fd34a_2.tar.bz2"
                    .parse().unwrap()
            ),
            Some("osx-arm64")
        );
        assert_eq!(
            derive_subdir_from_url(
                &"https://conda.anaconda.org/conda-forge/win-64/package-1.0.0-0.tar.bz2"
                    .parse()
                    .unwrap()
            ),
            Some("win-64")
        );
        assert_eq!(
            derive_subdir_from_url(
                &"https://conda.anaconda.org/conda-forge/noarch/package-1.0.0-0.tar.bz2"
                    .parse()
                    .unwrap()
            ),
            Some("noarch")
        );
    }

    #[test]
    fn test_channel_from_url() {
        assert_eq!(derive_channel_from_url(&Url::parse("https://conda.anaconda.org/conda-forge/osx-64/python-3.11.0-h4150a38_1_cpython.conda").unwrap()), Some(Url::parse("https://conda.anaconda.org/conda-forge").unwrap().into()));
        assert_eq!(
            derive_channel_from_url(
                &Url::parse(
                    "file:///C:/Users/someone/AppData/Local/Temp/.tmpsasJ7b/noarch/foo-1-0.conda"
                )
                .unwrap()
            ),
            Some(
                Url::parse("file:///C:/Users/someone/AppData/Local/Temp/.tmpsasJ7b")
                    .unwrap()
                    .into()
            )
        );
        assert_eq!(
            derive_channel_from_url(
                &Url::parse("https://repo.anaconda.com/pkgs/main/linux-64/package-1.0.0-0.tar.bz2")
                    .unwrap()
            ),
            Some(
                Url::parse("https://repo.anaconda.com/pkgs/main")
                    .unwrap()
                    .into()
            )
        );
        assert_eq!(
            derive_channel_from_url(
                &Url::parse("https://repo.anaconda.com/pkgs/free/noarch/package-1.0.0-0.tar.bz2")
                    .unwrap()
            ),
            Some(
                Url::parse("https://repo.anaconda.com/pkgs/free")
                    .unwrap()
                    .into()
            )
        );
    }

    #[test]
    fn test_derive_noarch_type() {
        assert_eq!(
            derive_noarch_type("noarch", "py313hb6a6212_1"),
            NoArchType::python()
        );
        assert_eq!(
            derive_noarch_type("noarch", "313hb6a6212_1"),
            NoArchType::generic()
        );
        assert_eq!(
            derive_noarch_type("linux-aarch64", "313hb6a6212_1"),
            NoArchType::none()
        );
        assert_eq!(
            derive_noarch_type("noarch", "generic_0"),
            NoArchType::generic()
        );
        assert_eq!(derive_noarch_type("win-64", "py_0"), NoArchType::none());
    }

    #[test]
    fn test_derive_platform_and_arch() {
        assert_eq!(
            derive_arch_and_platform("linux-aarch64"),
            (Some("aarch64".to_string()), Some("linux".to_string()))
        );
        assert_eq!(
            derive_arch_and_platform("osx-arm64"),
            (Some("arm64".to_string()), Some("osx".to_string()))
        );
        assert_eq!(
            derive_arch_and_platform("win-64"),
            (Some("x86_64".to_string()), Some("win".to_string()))
        );
        assert_eq!(derive_arch_and_platform("noarch"), (None, None));
        assert_eq!(derive_arch_and_platform("unknown"), (None, None));
        assert_eq!(derive_arch_and_platform("win-64-2"), (None, None));
        assert_eq!(
            derive_arch_and_platform("emscripten-wasm32"),
            (Some("wasm32".to_string()), Some("emscripten".to_string()))
        );
        assert_eq!(
            derive_arch_and_platform("wasi-wasm32"),
            (Some("wasm32".to_string()), Some("wasi".to_string()))
        );
    }

    #[rstest]
    #[case(
        1,
        "https://conda.anaconda.org/conda-forge/linux-aarch64/qt-5.15.8-h803f02a_0.conda"
    )]
    #[case(
        2,
        "https://repo.anaconda.com/pkgs/main/linux-64/package-1.0.0-0.tar.bz2"
    )]
    #[case(
        3,
        "https://conda.anaconda.org/conda-forge/osx-arm64/x264-1!164.3095-h57fd34a_2.tar.bz2"
    )]
    #[case(4, "file:///packages/win-64/x264-1!164.3095-h57fd34a_2.tar.bz2")]
    #[case(5, "../source/win-64/package-1.0.0-0.tar.bz2")]
    #[case(6, "../source/win-64")]
    #[case(7, "../source/win-64/")]
    fn test_location_derived_fields(#[case] idx: usize, #[case] location_str: &str) {
        let location = UrlOrPath::from_str(location_str).unwrap();
        let fields = LocationDerivedFields::new(&location);
        insta::assert_debug_snapshot!(
            format!("location_derived_fields-{idx}"),
            fields,
            location_str
        );
    }
}
