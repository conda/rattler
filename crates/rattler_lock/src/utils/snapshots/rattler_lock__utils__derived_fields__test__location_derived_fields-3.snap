---
source: crates/rattler_lock/src/utils/derived_fields.rs
expression: "https://conda.anaconda.org/conda-forge/osx-arm64/x264-1!164.3095-h57fd34a_2.tar.bz2"
---
LocationDerivedFields {
    file_name: Some(
        "x264-1!164.3095-h57fd34a_2.tar.bz2",
    ),
    name: Some(
        "x264",
    ),
    version: Some(
        "1!164.3095",
    ),
    build: Some(
        "h57fd34a_2",
    ),
    subdir: Some(
        "osx-arm64",
    ),
    channel: Some(
        "https://conda.anaconda.org/conda-forge/",
    ),
}
